feat: Library Home - Paste Content [FC-0059]

yusuf-musleh opened the pull request:
## Description

This PR adds the ability to paste blocks from the clipboard into a v2 Library. It adds a new button "Paste From Clipboard" that appears the user's clipboard contains a block that can be pasted into the library. In the case where the clipboard contains a unit (which cannot be pasted in a library), then the "Paste From Clipboard" button does not appear.

<img width="926" alt="Screenshot 2024-08-08 at 7 29 05 AM" src="https://github.com/user-attachments/assets/081077f3-49bc-44f1-a39d-175aaaaab37b">

## Supporting information

Related Ticket:
- https://github.com/openedx/frontend-app-course-authoring/issues/1176
- Depends on https://github.com/openedx/edx-platform/pull/35199/

## Testing instructions

1. Run you devstack on the branch in https://github.com/openedx/edx-platform/pull/35199/
1. Run the frontend on this branch
1. Make sure you have a v2 library created, create one if not, and add a bunch of components to it
1. Navigate to the v2 library, and click on the 3-dot menu of a component to copy it to clipboard
1. Then click on the "+ New" button you will notice the "Paste From Clipboard" button appear
1. Click on the "Paste From Clipboard" button and confirm that the copied component is added to the library successfully and matches the content that was copied to clipboard
1. Try copying from course units/blocks, navigate to a course that has units/blocks:
    1. Copy a unit from the course
    3. Navigate to a v2 library, and click on the "+ New" button
    4. Confirm that the "Paste From Clipboard" button does **not** appear
    5. Navigate back to the course, and copy a block from inside a unit
    6. Navigate back to the v2 library
    7. Confirm that the "Paste From Clipboard" button now appears
1. Click on the "Paste From Clipboard" button and confirm that the copied block is added to the library successfully and matches the content of the block in the clipboard
1. Confirm that a toast message is shown while pasting and when successfully pasted.
1. Confirm that the "Paste From Clipboard" button appears/disappears when blocks/units are copied to clipboard from other browser tabs/windows

---
Private-ref: [FAL-3778](https://tasks.opencraft.com/browse/FAL-3778)

openedx-webhooks commented:
Thanks for the pull request, @yusuf-musleh!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->


#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/2u-tnl`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

bradenmacdonald commented:
While you're at this, I noticed that all our icon `es5/*.js` files all start with this line of Babel boilerplate that's not needed anymore, but significantly increases all our icon file sizes:

```js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
```

I found you can fix it with this change, which significantly reduces the total size of the `es5` folder (though it's actual effect on MFE build size varies a lot depending on the MFE's own babel config; the effect on the gzipped size is ~none in any case).
<img width="631" alt="Screenshot 2024-07-12 at 12 16 01 PM" src="https://github.com/user-attachments/assets/1ea95a4f-52b6-4e36-be8e-f0bd213558bf">

bradenmacdonald commented:
While you're at this, I noticed that all our icon `es5/*.js` files all start with this line of Babel boilerplate that's not needed anymore, but significantly increases all our icon file sizes:

```js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
```

I found you can fix it with this change, which significantly reduces the total size of the `es5` folder (though it's actual effect on MFE build size varies a lot depending on the MFE's own babel config; the effect on the gzipped size is ~none in any case).

![Screenshot 2024-07-29 at 10 58 16 AM](https://github.com/user-attachments/assets/a83c46f5-e9b0-4d0f-a530-c0d6ef23c968)

ChrisChV commented:
@yusuf-musleh https://github.com/openedx/frontend-app-course-authoring/pull/1138 and https://github.com/openedx/frontend-app-course-authoring/pull/1197 are Done, could you update your branch and test with copy from a library?

yusuf-musleh commented:
@ChrisChV Thanks for the review! I've updated my branch and tested with copy from library and pasting, it working fine. However, it was missing broadcasting the clipboard data to update whether the "Paste from Clipboard" button appears without refreshing, so I added it here: [44cb93c](https://github.com/openedx/frontend-app-course-authoring/pull/1187/commits/44cb93cc24d19963d5e0c200722c502180164c4c), I'm working on fixing the broken tests.

ChrisChV commented:
@yusuf-musleh  The "Paste From Clipboard" doesn't dissapear on paste. https://www.loom.com/share/74030a6a5aae437e9b866971c3a434af?sid=fe67f250-6363-4650-a026-6c717fbe9019

bradenmacdonald commented:
@ChrisChV Isn't that expected behavior? Pasting something doesn't remove it from your clipboard, and you are able to paste it multiple times (into the same library or into several different libraries/courses) if you want to.

ChrisChV commented:
> Isn't that expected behavior?

I don't see anything mentioned. I find it odd that I can paste the same component multiple times. But I'm also fine with copying not removing the item from the clipboard.

ChrisChV reviewed (approved):
@yusuf-musleh  Looks good :+1:
- [X] I tested this: I copy multiple components from a course and from a library
- [X] I read through the code and considered the security, stability and performance implications of the changes.
- [X] I tested that the UI can be used with a keyboard only (tab order, keyboard controls).
- [X] Includes tests for bugfixes and/or features added.
- [X] Includes documentation

yusuf-musleh commented:
@ChrisChV

> I don't see anything mentioned. I find it odd that I can paste the same component multiple times. But I'm also fine with copying not removing the item from the clipboard.

It is the same behavior we have when pasting a unit inside a course, the "Paste Unit" remains there and you can continue pasting the same unit from the clipboard.

bradenmacdonald commented:
> I find it odd that I can paste the same component multiple times.

This clipboard is supposed to work the same way as the "regular" clipboard on your computer. If you copy something, e.g. text in your browser right now or code in VS Code or anything, you can always paste it multiple times.

ChrisChV commented:
> This clipboard is supposed to work the same way as the "regular" clipboard on your computer. If you copy something, e.g. text in your browser right now or code in VS Code or anything, you can always paste it multiple times.

You're right. Thanks :+1:

bradenmacdonald commented on src/generic/clipboard/hooks/useCopyToClipboard.js:
```diff
@@ -30,6 +34,22 @@ const useCopyToClipboard = (canEdit = true) => {
+    };
+
+    fetchInitialClipboardData();
+  }, [dispatch]);
```
As a general approach, I would prefer we replace the redux code with something that uses React Query, rather than finding a way to make the redux code work better. But this is fine for now.

See recently updated frontend recommendations: https://docs.openedx.org/projects/openedx-proposals/en/latest/best-practices/oep-0067-bp-tools-and-technology.html#frontend-technology-selection

yusuf-musleh replied:
Agreed, makes sense, I just wasn't sure if I wanted to do the full refactor to React Query as part of this task/PR, as opposed to a separate PR/task, since it's not specific to pasting in the library.

openedx-webhooks commented:
@yusuf-musleh 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/frontend-app-course-authoring&entry.867055334=https://github.com/openedx/frontend-app-course-authoring/pull/1187&entry.1484655318=https://github.com/yusuf-musleh&entry.752974735=2024-07-29+12:03&entry.1917517419=2024-08-15+17:03&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
